### PR TITLE
fix(macos/input): incorrect mouse input for non-main display

### PR DESCRIPTION
**Pull request by TimmyOVO**
## Description

fix: abs mouse input not working in macOS
fix: wrong abs mouse coordinates when capturing non-main display in macOS
fix: wrong mouse coordinates correction when capturing non-main display in macOS

## Type of Change
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Dependency update (updates to dependencies)
- [ ] Documentation update (changes to documentation)
- [ ] Repository update (changes to repository files, e.g. `.github/...`)

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated the in code docstring/documentation-blocks for new or existing methods/components

## Branch Updates
LizardByte requires that branches be up-to-date before merging. This means that after any PR is merged, this branch
must be updated before it can be merged. You must also
[Allow edits from maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I want maintainers to keep my branch updated

### CODE REVIEW

**Comment by ReenigneArcher** [comment]
I don't have the ability to test these changes. Ideally unit tests would be added around the changed code, but I understand that's not always the easiest thing to do.

Can you explain how you tested your changes?

File: src/platform/macos/display.mm
Comment (TimmyOVO) [inline src/platform/macos/display.mm:1]: In this [commit](https://github.com/LizardByte/Sunshine/commit/a4d9ee3fa4630bc0e7fdff230e923829c1d06015#diff-1518590f5c5028e6a34dd695918cfc679e97e1452dc0dc814718f250d93795deR481), there is a change that would require both env_width and env_height to have non-zero values, but these values are not properly set in that commit, which could break absolute mouse input on macOS.

File: src/platform/macos/input.cpp
Comment (TimmyOVO) [inline src/platform/macos/input.cpp:1]: Before we support change the display we need to capture in macOS, the location we get here according to [Apple Document](https://developer.apple.com/documentation/coregraphics/1455788-cgeventgetlocation?language=objc) should be `The current location of the specified mouse event in global display coordinates`, we can not just simply assuming the display's bounds  to `[0,display_width]` and `[0,display_height]`. So we need to get the correspond display bounds for display we are capturing using [`CGDisplayBounds`](https://developer.apple.com/documentation/coregraphics/1456395-cgdisplaybounds?language=objc) which return `The bounds of the display, expressed as a rectangle in the global display coordinate space (relative to the upper-left corner of the main display).`.

File: src/platform/macos/input.cpp [hunk 1]
```diff
@@ -379,10 +383,15 @@ const KeyCodeMap kKeyCodesMap[] = {
...
   abs_mouse(input_t &input, const touch_port_t &touch_port, float x, float y) {
-    auto scaling = ((macos_input_t *) input.get())->displayScaling;
+    auto macos_input = static_cast<macos_input_t *>(input.get());
+    auto scaling = macos_input->displayScaling;
+    auto display = macos_input->display;
 
     CGPoint location = CGPointMake(x * scaling, y * scaling);
-
+    CGRect display_bounds = CGDisplayBounds(display);
+    // in order to get the correct mouse location for capturing display , we need to add the display bounds to the location
```
Comment (TimmyOVO) [inline src/platform/macos/input.cpp:392]: The location we received are relative to the display we are capturing ,so we get the display bounds using [CGDisplayBounds](https://developer.apple.com/documentation/coregraphics/1456395-cgdisplaybounds?language=objc) then add them up to get the right input coordinates.

**Review by TimmyOVO — COMMENTED** [review]
That's all i done here in order to get correct input coordinates after we support specific display capture.

File: src/platform/macos/input.cpp [hunk 2]
```diff
@@ -328,15 +328,19 @@ const KeyCodeMap kKeyCodesMap[] = {
...
 
-    if (location.y < 0)
-      location.y = 0;
-    if (location.y >= (double) CGDisplayPixelsHigh(display))
-      location.y = (double) CGDisplayPixelsHigh(display) - 1;
+    // limit mouse to current display bounds
+    if (location.x < display_bounds.origin.x)
+      location.x = display_bounds.origin.x;
+    if (location.x >= display_bounds.origin.x + display_bounds.size.width)
+      location.x = display_bounds.origin.x + display_bounds.size.width - 1;
```
Comment (cgutman) [inline src/platform/macos/input.cpp:338]: You should use `std::clamp()` instead.

Example:
```
location.x = std::clamp(location.x,  display_bounds.origin.x, display_bounds.origin.x + display_bounds.size.width - 1);
```

**Comment by BayLee4** [comment]
Hi, just wanted to report that I tested the changes (by forking the homebrew tap [repository](https://github.com/LizardByte/homebrew-homebrew) and then pointing it to the right branch/commit) and confirm the issue is now gone, everything[1] works perfectly 👍

[1]: Well except the CTRL + ALT + SHIFT + N shortcut which still doesn't hide the cursor in anything but the main display, but that behavior is already present upstream so that would be a different issue

**Pull request merged**